openedx/core/djangoapps/course_groups/tests/test_cohorts.py

```diff
@@ -258,6 +258,31 @@ def test_get_cohort(self):
             "other_user should be assigned to the default cohort"
         )
 
+    def test_get_cohort_preassigned_user(self):
+        """
+        When an email address is added to a cohort and a user signs up for the course with that email address,
+        the user should automatically be added to that cohort and not a random cohort.
+        """
+        course = modulestore().get_course(self.toy_course_key)
+        cohort = CohortFactory(course_id=course.id, name="TestCohort", users=[])
+        cohort2 = CohortFactory(course_id=course.id, name="RandomCohort", users=[])
+        config_course_cohorts(course, is_cohorted=True)
+
+        # Add email address to the cohort
+        (user, previous_cohort, prereg) = cohorts.add_user_to_cohort(cohort, "email@example.com")
+        self.assertEquals(
+            (user.email, previous_cohort, prereg),
+            ("email@example.com", None, True)
+        )
+
+        # Create user with this email address
+        user = UserFactory(username="test", email="email@example.com")
```

Comment:
Good test. It would also be nice to test that if you assign the same email addresses with multiple cohorts, the last assigned cohorts is the one that is respected.